src/network/networkd-link.c

```diff
@@ -2756,19 +2756,35 @@ int link_update(Link *link, sd_netlink_message *m) {
-
-                memcpy(link->mac.ether_addr_octet, mac.ether_addr_octet, ETH_ALEN);
-
-                log_link_debug(link, "Gained new MAC address: "
-                               "%02hhx:%02hhx:%02hhx:%02hhx:%02hhx:%02hhx",
-                               mac.ether_addr_octet[0],
-                               mac.ether_addr_octet[1],
-                               mac.ether_addr_octet[2],
-                               mac.ether_addr_octet[3],
-                               mac.ether_addr_octet[4],
-                               mac.ether_addr_octet[5]);
+        r = netlink_message_read_hw_addr(m, IFLA_ADDRESS, &hw_addr);
+        if (r >= 0 && (link->hw_addr.length != hw_addr.length ||
+                       memcmp(link->hw_addr.addr.bytes, hw_addr.addr.bytes, hw_addr.length) != 0)) {
+
+                memcpy(link->hw_addr.addr.bytes, hw_addr.addr.bytes, hw_addr.length);
+
+                if (link->iftype == ARPHRD_INFINIBAND)
+                        log_link_debug(link, "Gained new hardware address: "
+                                       "%02hhx:%02hhx:%02hhx:%02hhx:%02hhx:%02hhx:%02hhx:%02hhx:%02hhx:%02hhx"
+                                       ":%02hhx:%02hhx:%02hhx:%02hhx:%02hhx:%02hhx:%02hhx:%02hhx:%02hhx:%02hhx",
+                                       hw_addr.addr.infiniband[ 0], hw_addr.addr.infiniband[ 1],
+                                       hw_addr.addr.infiniband[ 2], hw_addr.addr.infiniband[ 3],
+                                       hw_addr.addr.infiniband[ 4], hw_addr.addr.infiniband[ 5],
+                                       hw_addr.addr.infiniband[ 6], hw_addr.addr.infiniband[ 7],
+                                       hw_addr.addr.infiniband[ 8], hw_addr.addr.infiniband[ 9],
+                                       hw_addr.addr.infiniband[10], hw_addr.addr.infiniband[11],
+                                       hw_addr.addr.infiniband[12], hw_addr.addr.infiniband[13],
+                                       hw_addr.addr.infiniband[14], hw_addr.addr.infiniband[15],
+                                       hw_addr.addr.infiniband[16], hw_addr.addr.infiniband[17],
+                                       hw_addr.addr.infiniband[18], hw_addr.addr.infiniband[19]);
+                else
+                        log_link_debug(link, "Gained new MAC address: "
+                                       "%02hhx:%02hhx:%02hhx:%02hhx:%02hhx:%02hhx",
+                                       hw_addr.addr.ether.ether_addr_octet[0],
+                                       hw_addr.addr.ether.ether_addr_octet[1],
+                                       hw_addr.addr.ether.ether_addr_octet[2],
+                                       hw_addr.addr.ether.ether_addr_octet[3],
+                                       hw_addr.addr.ether.ether_addr_octet[4],
+                                       hw_addr.addr.ether.ether_addr_octet[5]);
```

Comment:
maybe add hw_addr_format() or so, that does this formatting?

Reply:
Yeah, something that dynamically prints it is definitely needed.
I'll see if I can come up with something nicer.